### PR TITLE
Admin ui 2409239

### PR DESCRIPTION
**Pull request by ndf**
Create a Rules UI skeleton & allow to manage components
https://www.drupal.org/node/2409239

### CODE REVIEW

File: modules/rules_ui/rules_ui.links.action.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Action links for Rules UI.
+
+# Rules reaction related actions.
+rules_reaction_add_action:
+  route_name: stub.rules_reaction.add_form
+  title: 'Add new rules reaction'
```
Comment (fago) [inline modules/rules_ui/rules_ui.links.action.yml]: Let's call this "Add reaction rule"

File: modules/rules_ui/rules_ui.links.action.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  appears_on:
+    - stub.rules_reaction.collection
+
+# Rules component related actions.
+rules_component_add_action:
+  route_name: entity.rules_component.add_form
+  title: 'Add new rules component'
```
Comment (fago) [inline modules/rules_ui/rules_ui.links.action.yml]: Let's call this "Add component"

File: modules/rules_ui/rules_ui.links.menu.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+# rules.ui.overview is declared as parent.
+rules.ui.overview:
+  title: 'Rules'
+  route_name: rules.ui.overview
+  parent: system.admin_config_workflow
+  description: 'Rules administration overview.'
+
+# Menu links shown on rules.ui.overview .
+rules.ui.reactions:
+  title: 'Reactions'
```
Comment (fago) [inline modules/rules_ui/rules_ui.links.menu.yml]: Reaction rules

File: modules/rules_ui/rules_ui.links.task.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Menu tasks for Rules UI.
+
+rules.ui.overview:
+  title: 'Overview'
+  route_name: rules.ui.overview
+  base_route: rules.ui.overview
+  weight: 0
+
+rules.ui.reactions:
+  title: 'Reactions'
```
Comment (fago) [inline modules/rules_ui/rules_ui.links.task.yml]: Reaction rules

File: modules/rules_ui/rules_ui.routing.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  requirements:
+    _permission: 'administer rules'
+
+
+### Rules Reactions
+stub.rules_reaction.collection:
+  path: '/admin/config/workflow/rules/reactions'
+  defaults:
+    _controller: 'Drupal\rules\Stub\RulesReactionStub::notImplemented'
+    _title: 'Rules Reactions'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Reaction rules

File: modules/rules_ui/rules_ui.routing.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  requirements:
+    _permission: 'administer rules+administer rules reactions'
+
+stub.rules_reaction.add_form:
+  path: '/admin/config/workflow/rules/reactions/add'
+  defaults:
+    _controller: 'Drupal\rules\Stub\RulesReactionStub::notImplemented'
+    _title: 'Rules add Reaction'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Add reaction rule

File: modules/rules_ui/rules_ui.routing.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  requirements:
+    _permission: 'administer rules+administer rules reactions'
+
+stub.rules_reaction.edit_form:
+  path: '/admin/config/workflow/rules/reactions/edit/{rules_reaction}'
+  defaults:
+    _controller: 'Drupal\rules\Stub\RulesReactionStub::notImplemented'
+    _title: 'Rules edit Reaction'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Edit reaction rule

File: modules/rules_ui/rules_ui.routing.yml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  requirements:
+    _permission: 'administer rules+administer rules reactions'
+
+stub.rules_reaction.delete_form:
+  path: '/admin/config/workflow/rules/reactions/delete/{rules_reaction}'
+  defaults:
+    _controller: 'Drupal\rules\Stub\RulesReactionStub::notImplemented'
+    _title: 'Rules delete Reaction'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Delete reaction rule

File: modules/rules_ui/rules_ui.routing.yml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    _permission: 'administer rules+administer rules reactions'
+
+
+### Rules Components
+entity.rules_component.collection:
+  path: '/admin/config/workflow/rules/components'
+  defaults:
+  # Calls the list controller, defined in the annotation of the rules component entity.
+    _entity_list: 'rules_component'
+    _title: 'Rules Components'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Components

File: modules/rules_ui/rules_ui.routing.yml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    _entity_form: 'rules_component.add'
+    _title: 'Rules add Component'
+  requirements:
+    _permission: 'administer rules+administer rules components'
+
+entity.rules_component.edit_form:
+  path: '/admin/config/workflow/rules/components/edit/{rules_component}'
+  defaults:
+    _entity_form: 'rules_component.edit'
+    _title: 'Rules edit Component'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Edit components

File: modules/rules_ui/rules_ui.routing.yml [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  requirements:
+    _permission: 'administer rules+administer rules components'
+
+entity.rules_component.delete_form:
+  path: '/admin/config/workflow/rules/components/delete/{rules_component}'
+  defaults:
+    _entity_form: 'rules_component.delete'
+    _title: 'Rules delete Component'
```
Comment (fago) [inline modules/rules_ui/rules_ui.routing.yml]: Delete component

File: rules.permissions.yml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+# Most of the time administrators will be able to adminiter both reactions,
```
Comment (fago) [inline rules.permissions.yml]: administer

File: src/Entity/Controller/RulesComponentListBuilder.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Entity\Controller\RulesComponentController.
```
Comment (fago) [inline src/Entity/Controller/RulesComponentListBuilder.php]: outdated

File: src/Entity/Controller/RulesComponentListBuilder.php [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+namespace Drupal\rules\Entity\Controller;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Config\Entity\ConfigEntityListBuilder;
+use Drupal\Core\Url;
+
+/**
+ * Provides a list controller for rules controller entity.
```
Comment (fago) [inline src/Entity/Controller/RulesComponentListBuilder.php]: it'S a handler

File: src/Entity/RulesComponent.php [hunk 15]
```diff
@@ -142,6 +154,20 @@ public function label() {
...
+   * Returns the description.
+   */
+  public function getDescription() {
+    return $this->description;
+  }
+
+  /**
+   * Returns the tag.
+   */
+  public function getTag() {
```
Comment (fago) [inline src/Entity/RulesComponent.php:166]: We actually have multiple tags. I think that's broken right now, but let's do already a proper method names -> getTags() which returns string[].

File: src/Entity/RulesComponentAddForm.php [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+ * Provides a form for adding rules components.
+ */
+class RulesComponentAddForm extends RulesComponentFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function actions(array $form, FormStateInterface $form_state) {
+    $actions = parent::actions($form, $form_state);
+    $actions['submit']['#value'] = $this->t('Add new rules component');
```
Comment (fago) [inline src/Entity/RulesComponentAddForm.php]: Add component

File: src/Entity/RulesComponentAddForm.php [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    return $actions;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function save(array $form, FormStateInterface $form_state) {
+    parent::save($form, $form_state);
+
+    drupal_set_message($this->t('Rules component %label was created', array('%label' => $this->entity->label())));
```
Comment (fago) [inline src/Entity/RulesComponentAddForm.php]: Please take over the message from d7.

File: src/Entity/RulesComponentDeleteForm.php [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Builds a form to delete an action.
+ */
+class RulesComponentDeleteForm extends EntityConfirmFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQuestion() {
+    return $this->t('Are you sure you want to delete the rules component %label?', array('%label' => $this->entity->label()));
```
Comment (fago) [inline src/Entity/RulesComponentDeleteForm.php]: remove "rules"

File: src/Entity/RulesComponentDeleteForm.php [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $this->entity->delete();
+
+    $this->logger('user')->notice('Deleted rules component %label)', array('%label' => $this->entity->label()));
+    drupal_set_message($this->t('Rules component %label was deleted', array('%label' => $this->entity->label())));
```
Comment (fago) [inline src/Entity/RulesComponentDeleteForm.php]: Missing trailing point. "Component %label has been deleted.".

File: src/Entity/RulesComponentEditForm.php [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+ * Provides a form to edit rules components.
+ */
+class RulesComponentEditForm extends RulesComponentFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function actions(array $form, FormStateInterface $form_state) {
+    $actions = parent::actions($form, $form_state);
+    $actions['submit']['#value'] = $this->t('Save rules component');
```
Comment (fago) [inline src/Entity/RulesComponentEditForm.php]: Just save?

File: src/Entity/RulesComponentEditForm.php [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    return $actions;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function save(array $form, FormStateInterface $form_state) {
+    parent::save($form, $form_state);
+    drupal_set_message($this->t('Rules component %label was saved', array('%label' => $this->entity->label())));
```
Comment (fago) [inline src/Entity/RulesComponentEditForm.php]: has been + trailing point + remove "rules".

File: src/Entity/RulesComponentFormBase.php [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+abstract class RulesComponentFormBase extends EntityForm {
+
+  /**
+   * The rules component entity.
+   *
+   * @var \Drupal\rules\Entity\RulesComponent
+   */
+  protected $entity;
+
+  protected $storage = NULL;
```
Comment (fago) [inline src/Entity/RulesComponentFormBase.php:32]: misses comment.

File: src/Entity/RulesComponentFormBase.php [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Constructs a new ActionAddForm.
```
Comment (fago) [inline src/Entity/RulesComponentFormBase.php:35]: Constructs the object

File: modules/rules_ui/rules_ui.links.action.yml (hunk 1 above)
Comment (ndf) [inline modules/rules_ui/rules_ui.links.action.yml]: done

File: modules/rules_ui/rules_ui.links.action.yml (hunk 2 above)
Comment (ndf) [inline modules/rules_ui/rules_ui.links.action.yml]: done

File: modules/rules_ui/rules_ui.routing.yml (hunk 10 above)
Comment (ndf) [inline modules/rules_ui/rules_ui.routing.yml]: Made it 'Edit component'  (singular)

File: src/Entity/RulesComponentAddForm.php (hunk 17 above)
Comment (ndf) [inline src/Entity/RulesComponentAddForm.php]: Not all actions in rules d7 have a message. For now we follow core (block and node) with this pattern: 
%type %label has been <action>.

Component %label has been deleted.

File: src/Entity/RulesComponentFormBase.php (hunk 22 above)
Comment (ndf) [inline src/Entity/RulesComponentFormBase.php:32]:
```

  protected $entity;

  /**
   * The RulesComponent storage.
   *
   * @var \Drupal\Core\Entity\EntityStorageInterface
   */
  protected $storage = NULL;

  /**

```

Pushed this comment, source 'action' module

File: src/Entity/RulesComponent.php (hunk 15 above)
Comment (ndf) [inline src/Entity/RulesComponent.php:166]: Don't know howto do this.
Klausi commented on protected $tag (in class RulesComponent line ~70)

```
  /**
   * The "tags" of a Rules component.
   *
   * The tags are stored as a single string, though it is used as multiple tags
   * for example in the rules overview.
   *
   * @var string
   */
  protected $tag = '';
```

Prefer to move to follow up ticket.

**Comment by fago** [comment]
Merged manually, thanks!

**Pull request closed without merging**